Start of Bluetooth Handsfree rework

lkernan opened the pull request:
This reworks the BonovoBluetooth handsfree phone app in the following ways:
- Replaces the old Boolean method of switching views with one based on enum values.  It should be easier to maintain in future.
- Adds a full screen phone ringing screen without the dial buttons. 
  It's not very pretty yet, i've gone for function over looks so far.  Bigger text and buttons make it easier to answer a call when driving.
- Adds a button on the in call screen to show or hide the dialing buttons.  Allows for phone systems where you have to press numbers in a call.  
  The button to go back from the dialer to the call is the back facing arrow in the top left corner.

Future work will include adding in call buttons for holding and resuming calls, handling call waiting and conference calls, etc.
There shouldn't be any loss of functionality over the original system at this stage.

lkernan committed:
Full screen Bluetooth answer screen

When a call comes in via Bonovo Bluetooth, hide the dialling buttons and
show a full screen answering layout.

lkernan committed:
Build fix

Remove remaining code references to margins that have been removed

lkernan committed:
Hide the right panel

There are several nested layouts in the dialer, I was hiding the wrong
one.

lkernan committed:
Rework handsfree layout switching code

Replace the old method of passing boolean values to select the view with
a more descriptive and extensible method.  Also update the layout of the
ringing page to something a little more user friendly.

lkernan committed:
Allow switching between incall and dialpad views

Add a button to switch view between in call view and dialpad view.
Several more buttons will be added to in call screen for hold / resume
call, switch audio source, etc.

lkernan committed:
Don't cut bottom off caller name text

lkernan committed:
Don't call a nonexistant service

It's possible to show the dialer when MyBluetoothService  is not
available.
We should handle that without crashing by just showing the dialpad.

lkernan committed:
Current Progress

Push my current progress to my github account.  Microphone mute works,
caller contact photos work sometimes.

There is the start of a call waiting UI but no backend support yet.
Newsmy didn't get around to adding those commands yet so we'll have to
do it ourselves.

jamesgol commented on BonovoBluetooth/src/com/bonovo/bluetooth/BonovoBluetoothHandfree.java:
```diff
@@ -359,49 +387,59 @@ protected void onResume() {
 			}
 		}
 		mDialpadPressCount = 0;
+		
```
Can we try not to make so many whitespace changes like these?  In the future if there is another source leak it just makes it a bit harder to do a comparison of changes.

lkernan replied:
Sorry, i'll try to remember not to do that again.
Had a problem tonight with the Chinese text in one of the CPP files getting garbled.  Sorry for the noise but I'm pretty sure everything is back in order now.

lkernan replied:
I need to test my latest changes in the car yet, but everything compiles and runs in the house.  

All the JNI code should be in place for call waiting, getting Bluetooth media information (track name, artist, current position, etc) and sending custom commands to the phone via SPP.

In the next day or two i'll try to get the call waiting service code connected to the UI and it should be looking pretty good.

lkernan replied:
Ok, so I got a chance to test in the car as I drove to work this morning, and it crashes as soon as the phone connects.. 
I'll move the computer out to the garage tonight and sort it out. I'm jealous of you guys with more than one headunit, testing would be a lot easier inside :-)

jamesgol replied:
I actually did a test of the code earlier today and made a couple outgoing calls and it didn't crash for me.

jamesgol commented:
I haven't even tested any bluetooth calling on my setup yet so I can't test if it works, but on a quick glance everything looks good.  I think it's great to have someone else interested in working on this.

lkernan committed:
Mainly back end AT command changes

Add default contact photo if one is not found.
Add comments for Bluetooth AT commands.
Add additional commands to JNI backend code for call waiting, AVRCP
media information, SPP support and misc other small features

lkernan committed:
Revert "Mainly back end AT command changes"

This reverts commit 1d0cdb024827e0b02f1eb36a44e9960365f64f69.

lkernan committed:
Mainly back end AT command changes

Second try.  I reverted because the chinese writing in
com_bonovo_bluetooth_thread.cpp got screwed up but then everything
disappeared from my local copy when I reverted.  At this point i'm going
to put it back and the chinese comments are the least of my worries.

lkernan committed:
Fix chinese text in com_bonovo_bluetooth_thread.cpp

Not sure how I screwed that up but I think it's back as it was before.
Also added English translations.

lkernan committed:
Fix for signal strength and battery level response

According to the manual, these return a value between 0 and 5, so param only need return one character. 
For some reason it was appending a newline which caused problems further on.

lkernan committed:
Fixes for contact images

Additional sanity check on the service side for phone battery and signal
levels. They are now being reported correctly from my iPhone at least.

Contacts image view fix. It helps to inflate the layout before trying to
reference it's content.
Also change to using functions that don't decode the bitmap in the UI
thread.

lkernan committed:
Switch to AUDIOFOCUS_GAIN_TRANSIENT

Time for me to stop overthinking and listen to some good advise :-)
Set Bluetooth to use AudioManager.AUDIOFOCUS_GAIN_TRANSIENT

jamesgol commented:
This looks like it's shaping up nicely.  Can you post an APK the the forum for people to test as you keep making changes?  I'll try to test some of the stuff myself also.  When you are changes are completed we can merge this in and put out another release.

lkernan commented:
Yes, we're getting close to a point where I can put an APK out.  I'd just like to get the call waiting stuff finished (hopefully tonight).  
I've just about finished writing the service side intent broadcast and receivers, just need to hook them up to the UI side and test.

jamesgol commented:
Great!  We are going to split up this repository into the individual elements and keep the original intact to track past and future source leaks.  I think after your changes are finished we can commit them and then do the split up.  I think it will be a better long term strategy to have each of them split into their own repository.

lkernan commented:
One question.  I tried building just an APK file by doing "make BonovoBluetooth". 
I got an APK file in the out folder which claimed to have installed, but doesn't seem to have, it was still the previous version.  Up to now i've been building update ZIP files and that works.

Am i doing something wrong or is it just not possible to update the built in apps in that method?

I'm just doing a few cleanups on the code now, call waiting works and you can swap between the active and inactive calls fine. The UI isn't very pretty but that is quite easy to fix after the basics are tested a bit.

Zaphod-Beeblebrox commented:
Your make command should have worked fine, and it sounds like the new apk was placed in /out/system/app/ ??  If so, then it built right.

How did you push this to the android box?  You may not have luck 'installing it'.  You probably need to remove the original apk, and put the new one in /system/app/

Will require remounting /system.  I do this from the command line thus:

adb remount
adb push {newapk} /system/app/{newapk}
adb chmod 644 /system/app/{newapk}
adb reboot

lkernan commented:
Ok, that's what i did wrong, i put it on a USB stick and tried installing it via ES explorer.
Thanks for the tip.

Zaphod-Beeblebrox commented:
I just took a very quick look at a few commits, so please excuse me if I overlooked something.  But I see where you switched to AudioFocus (which I think will be a better solution), but I don't see where you are releasing the focus?  Need to do that on call end I think.

Zaphod-Beeblebrox commented:
Nevermind.  Just saw it .. there's a lot of code changes in here :)  I've not gone through much of it.

lkernan commented:
:-)  abandonAudioFocus();
Get called when phonestate switches to idle in a couple of places.

lkernan committed:
Call waiting

Add the required code to activate call waiting.
Also adjustments to layout to make swapping and ending call on hold a
little easier.
Missed one required Bluetooth command so it's been added in
com_bonovo_bluetooth_thread.cpp
Added a conference button to merge active and held calls into one.

Made contact image appear in a circle instead of square as requested.

lkernan commented:
Ok, i've just pushed my latest changes across.  I might hold off sending out an APK on the forums until tomorrow as it's getting late here and there are still a couple of minor issues i'd like to check and see if i can address.

Please feel free to give it a go, all the basic phone functions seem pretty stable now.  Call waiting has had only basic testing but works for me...

It seems a little touchy about connecting to the phone on first run after upgrading. Toggling the phone option in the settings screen gets it working again.

For some reason it's not closing when the call is over which i'd like to check.  Google Play music isn't continuing after the call so i'll check that too.  

Here is a link to the APK file on dropbox
https://www.dropbox.com/s/qavfo9f52efkae1/BonovoBluetooth.apk?dl=0 

If it crashes on a call the call will probably continue on without the UI.  In that case your phone will be able to control the call and end it ok.

lkernan commented:
Ok I think I found why music wasn't resuming.  It was passing null to the abandonAudioFocus call instead of the listener.  
I'll try a fresh build on my drive to work shortly and if it works ok i'll put that build out as a test version.

lkernan committed:
Fixes

Pass correct value to AbandonAudioFocus so hopefully music will restart
after the call
Shrink caller name text to half size in a conference so two lines fit.
Change colour of call waiting bar.

lkernan committed:
Claim exclusive audio focus while sleeping

Proposed change to BonovoHandle to cause music apps to pause while sleeping.
Completely untested at the moment. Written in the Githib web editor on my lunch break so hopefully it even just compiles

lkernan committed:
Fix typo

Fix an unintended typo that snuck in.

lkernan committed:
missing variable declaration

Declare result variable correctly

lkernan committed:
Build fix

Looking at what Newsmy were doing back in 1.0.9 reveals what I did wrong.

lkernan committed:
Working audio focus on sleep code

There was some old newsmy audio handling code in the handle service that confused me a bit.
After reading some of the documentation i've reworked the AudioFocusChangeListener and now it works.

Tested by hooking the Android box up to my PC and sending Screen-off and Screen-on intents
The service claims AudioManager.AUDIOFOCUS_GAIN_TRANSIENT_EXCLUSIVE and gives it up when the wake event occurs.

lkernan committed:
Merge pull request #1 from lkernan/musicsleeping

Claim exclusive audio focus while sleeping

lkernan commented:
How do you keep these separate?  The music sleeping change is good to go, but I expected to keep it separate from the bluetooth changes.

Zaphod-Beeblebrox commented:
Branches are the way to do it, but you probably shouldn't have merged your musicsleeping branch back to master.  Should issue a pull request directly from that branch to nu3001/master.

The other thing I would suggest is setting up another branch for your UI re-write code and pushing it to your github, but not a pull request to NU3001.

When you are happy with all of the work, you should learn how to rebase your commit down to one single commit, push that to your master branch, and then pull request to nu3001.

What ends up happening there is you push one single commit to the main repo, rather than all these little ones as you worked it out.
https://www.atlassian.com/git/tutorials/rewriting-history

lkernan committed:
Revert "Merge pull request #1 from lkernan/musicsleeping"

This reverts commit 684eafdecbe6e839065a8537be328eb21fe84abe, reversing
changes made to e805721e271e1b853b00f08a599a883bae4165ef.

lkernan commented:
Closed as i'm trying to clean up my github tree.  I've got a working branch and have reset master to nu3001 master.  From now on i'll work on the branch and push to my master.  Will create a new pull request from there later.
